odiglet/pkg/kube/runtime_details/workload_enabled_predicated.go

```diff
@@ -0,0 +1,105 @@
+package runtime_details
+
+import (
+	"github.com/odigos-io/odigos/common/consts"
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// this predicate is used for workload reconciler, and will only pass events
+// where the workload is changed to odigos instrumentation enabled.
+// This way, we don't need to run language detection downstream when unnecessary.
+// This also helps in managing race conditions, where we might re-add runtime details
+// which were just deleted by instrumentor controller and generate unnecessary noise
+// in the k8s eventual consistency model.
+type WorkloadEnabledPredicate struct {
+	predicate.Funcs
+}
+
+func (i *WorkloadEnabledPredicate) Create(e event.CreateEvent) bool {
+	enabled := isInstrumentationEnabled(e.Object)
+	// only handle new workloads if they start with instrumentation enabled
+	return enabled
+}
+
+func (i *WorkloadEnabledPredicate) Update(e event.UpdateEvent) bool {
+
+	if e.ObjectOld == nil {
+		return false
+	}
+	if e.ObjectNew == nil {
+		return false
+	}
+
+	// only run runtime inspection if the workload was not instrumented before
+	// and now it is.
+	oldEnabled := isInstrumentationEnabled(e.ObjectOld)
+	newEnabled := isInstrumentationEnabled(e.ObjectNew)
+	becameEnabled := !oldEnabled && newEnabled
+
+	// special case for deployments, where we also want to pass events if the
+	// deployment pods become available so language detection can run
+	oldDeployment, oldOk := e.ObjectOld.(*appsv1.Deployment)
+	newDeployment, newOk := e.ObjectNew.(*appsv1.Deployment)
```

Comment:
Just a suggestion:
* get kind of old object
* get kind of new object
* if different kinds return false
* else: switch case on kind
  *for each case cast to appropriate typ
check the condition after casting